### PR TITLE
RC-27 add searchHandler to find similar patients

### PR DESCRIPTION
**Pull request by AdamGrzybkowski**
https://issues.openmrs.org/browse/RC-27

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreServiceImpl.java [hunk 1]
```diff
@@ -99,7 +99,7 @@ public void setPatientService(PatientService patientService) {
 		this.patientService = patientService;
 	}
 
-    public void setLocationService(LocationService locationService) {
+	public void setLocationService(LocationService locationService) {
         this.locationService = locationService;
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreServiceImpl.java]: The formatting here does not look correct. :)

**Comment by dkayiwa** [comment]
@AdamGrzybkowski can we also add some unit tests?

File: omod/src/main/java/org/openmrs/module/registrationcore/web/rest/PatientsBySimilarPatientSearchHandler.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        String[] supportedFormats = { "yyyy-MM-dd'T'HH:mm:ss.SSSZ", "yyyy-MM-dd'T'HH:mm:ss.SSS",
+                "yyyy-MM-dd'T'HH:mm:ssZ", "yyyy-MM-dd'T'HH:mm:ssXXX", "yyyy-MM-dd'T'HH:mm:ss",
+                "yyyy-MM-dd HH:mm:ss", "yyyy-MM-dd" };
+        for (int i = 0; i < supportedFormats.length; i++) {
+            try {
+                Date date = DateTime.parse(birthdate, DateTimeFormat.forPattern(supportedFormats[i])).toDate();
+                return date;
+            }
+            catch (IllegalArgumentException ex) {
+                pex = ex;
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/registrationcore/web/rest/PatientsBySimilarPatientSearchHandler.java]: Why don't you just throw the exception from here and remove the pex variable all together?

File: omod/src/main/java/org/openmrs/module/registrationcore/web/rest/PatientsBySimilarPatientSearchHandler.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        person.setGender(requestContext.getParameter(GENDER));
+        person.setBirthdate(getDateFromString(requestContext.getParameter(BIRTHDATE)));
+
+        return new Patient(person);
+    }
+
+    private Date getDateFromString(String birthdate) {
+        IllegalArgumentException pex = null;
+        String[] supportedFormats = { "yyyy-MM-dd'T'HH:mm:ss.SSSZ", "yyyy-MM-dd'T'HH:mm:ss.SSS",
+                "yyyy-MM-dd'T'HH:mm:ssZ", "yyyy-MM-dd'T'HH:mm:ssXXX", "yyyy-MM-dd'T'HH:mm:ss",
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/registrationcore/web/rest/PatientsBySimilarPatientSearchHandler.java]: Are we just guessing the format? Or do we expect a specific one? :)

**Comment by AdamGrzybkowski** [comment]
@dkayiwa about unit tests: I wanted to do them. I even made test classes, but couldn't run them. I described my problem shortly in the comment section of the [issue](https://issues.openmrs.org/browse/RC-27). Do I have to add any xml files to make it work?

**Comment by dkayiwa** [comment]
@AdamGrzybkowski can i look at your test?

File: omod/src/main/java/org/openmrs/module/registrationcore/web/rest/PatientsBySimilarPatientSearchHandler.java (hunk 3 above)
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/registrationcore/web/rest/PatientsBySimilarPatientSearchHandler.java]: I like your honesty! :)

**Comment by AdamGrzybkowski** [comment]
@dkayiwa sure! I have added new commit :)

**Comment by dkayiwa** [comment]
@AdamGrzybkowski just add the event and idgen module dependencies to your omod's pom.

**Comment by AdamGrzybkowski** [comment]
Thanks @dkayiwa! PR updated

File: api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreServiceImpl.java (hunk 1 above)
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreServiceImpl.java]: @AdamGrzybkowski did you see the comment above?

File: omod/src/main/resources/webModuleApplicationContext.xml [hunk 4]
```diff
@@ -26,7 +26,7 @@
 	<!-- Annotation based controllers -->
 	<bean class="org.springframework.web.servlet.mvc.annotation.DefaultAnnotationHandlerMapping"/>
 	
-	<context:component-scan base-package="org.openmrs.module.registrationcore.web.controller" />
```
Comment (dkayiwa) [inline omod/src/main/resources/webModuleApplicationContext.xml]: Do you mind explaining a bit why you changed this from controller to rest?

File: pom.xml [hunk 5]
```diff
@@ -139,7 +139,6 @@
                 <groupId>joda-time</groupId>
                 <artifactId>joda-time</artifactId>
                 <version>2.2</version>
-                <scope>test</scope>
```
Comment (dkayiwa) [inline pom.xml]: Isn't this already provided. Did you get any errors before changing it?

File: omod/src/main/java/org/openmrs/module/registrationcore/web/rest/PatientsBySimilarPatientSearchHandler.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
+package org.openmrs.module.registrationcore.web.rest;
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/registrationcore/web/rest/PatientsBySimilarPatientSearchHandler.java]: Did you forget to add the license header?

File: api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreServiceImpl.java (hunk 1 above)
Comment (AdamGrzybkowski) [inline api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreServiceImpl.java]: Yes, I'll fixed that.

File: omod/src/main/resources/webModuleApplicationContext.xml (hunk 4 above)
Comment (AdamGrzybkowski) [inline omod/src/main/resources/webModuleApplicationContext.xml]: The @Component annotation in the searchHandler wasn't working without that and there is no package org.openmrs.module.registrationcore.web.controller in this project or am I missing something ?

File: pom.xml (hunk 5 above)
Comment (AdamGrzybkowski) [inline pom.xml]: Without deleting it I could only use this library in test classes.

File: omod/src/main/java/org/openmrs/module/registrationcore/web/rest/PatientsBySimilarPatientSearchHandler.java (hunk 6 above)
Comment (AdamGrzybkowski) [inline omod/src/main/java/org/openmrs/module/registrationcore/web/rest/PatientsBySimilarPatientSearchHandler.java]: done

File: omod/src/test/java/org/openmrs/module/registrationcore/web/rest/PatientsBySimilarPatientSearchHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
+package org.openmrs.module.registrationcore.web.rest;
```
Comment (AdamGrzybkowski) [inline omod/src/test/java/org/openmrs/module/registrationcore/web/rest/PatientsBySimilarPatientSearchHandlerTest.java]: done

File: pom.xml (hunk 5 above)
Comment (dkayiwa) [inline pom.xml]: Can i look at the error you got before making this change?

Comment (AdamGrzybkowski) [inline pom.xml]: You can see the error form Inteliji when trying to make project. 
Those libraries are only available in the test scope without this change.

screenshot from 2016-10-28 16 50 39

Comment (AdamGrzybkowski) [inline pom.xml]: Sorry @dkayiwa for not responding to your question. I have missed that one. 
I can confirm that core provides this library. 
pom.xml edited :)

File: omod/pom.xml [hunk 8]
```diff
@@ -21,9 +21,32 @@
...
+		<dependency>
+			<groupId>org.openmrs.module</groupId>
+			<artifactId>webservices.rest-omod-common</artifactId>
+			<version>2.16</version>
+			<scope>test</scope>
+			<classifier>tests</classifier>
+		</dependency>
 
+		<dependency>
+			<groupId>org.openmrs.module</groupId>
```
Comment (dkayiwa) [inline omod/pom.xml]: Shouldn't this have the provided scope?

File: omod/pom.xml [hunk 9]
```diff
@@ -62,6 +85,10 @@
 
 		<!-- End OpenMRS core -->
 
+		<dependency>
+			<groupId>joda-time</groupId>
```
Comment (dkayiwa) [inline omod/pom.xml]: Are these inheriting the provided scope from somewhere?

Comment (AdamGrzybkowski) [inline omod/pom.xml]: If I'm not wrong it should inherit from the main pom.xml. The scope is defined there as provided

**Pull request merged**